### PR TITLE
    koord-scheduler: support Reservation reserved CPU Cores

### LINKED ISSUE #566: [proposal] Reservation supports fine-grained CPU orchestration
**What is your proposal**:

Reservation should support fine-grained CPU orchestration. i.e. A reservation can allocate fine-grained CPU resources via our NodeNUMAResource plugin, and an owner pod can allocate reserved fine-grained CPU resources and get the same CPU binding result.

Furthermore, we expect a generic approach for other scheduling plugins to get compatible with the resource reservation. We've added some utility functions in #550 to facilitate this work.

**Why is this needed**:

Reservation is designed to reserve node resources for specific pods or workloads.  When an LSR/LSE pod is rescheduled, we hope to use resource reservation to ensure node resources. Nevertheless, it cannot reserve fine-grained CPU resources with our NodeNUMAResource plugin, because the reservation plugin does not identify the CPU allocation protocol in the pod annotations.

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by eahydra**
### Ⅰ. Describe what this PR does

Enhanced NodeNUMAResource scheduling plugin:
- Support reserving CPUSets through Reservation.  If there are reserved CPUSets through Reservation, the schedule Pod will allocate the reserved CPU Cores first, and then allocate from the remaining available CPU Cores on node.

### Ⅱ. Does this pull request fix one issue?

fix #566

### Ⅲ. Describe how to verify it

1. Create a Reservation to reserve some CPU cores on a given node with 16 cores
```bash
$ cat << EOF | kubectl create -f -
apiVersion: scheduling.koordinator.sh/v1alpha1
kind: Reservation
metadata:
  generateName: test-
spec:
  allocateOnce: false
  owners:
  - labelSelector:
      matchLabels:
        app: test-app
  template:
    metadata:
      name: curlimage-55859d896f-8dx9c
      namespace: default
      labels:
        koordinator.sh/qosClass: LSR
    spec:
      containers:
      - resources:
          limits:
            cpu: "4"
          requests:
            cpu: "4"
      priority: 9000
      priorityClassName: koord-prod
      nodeSelector:
        kubernetes.io/hostname: cn-beijing.10.0.3.245
      restartPolicy: Always
      schedulerName: koord-scheduler
  ttl: 30m0s
EOF
reservation.scheduling.koordinator.sh/test-gfr29 created
```
2. Verify Reservation got CPU Cores
```bash
$ kubectl get reservation -o yaml | grep cpuset
      scheduling.koordinator.sh/resource-status: '{"cpuset":"0-3"}'
```
3. Create a deployment with 3 replicas, per replica requests 2 Cores
```bash
$ cat << EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: test-app
  name: test-deploy
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: test-app
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        scheduling.koordinator.sh/resource-spec: '{"preferredCPUBindPolicy": "SpreadByPCPUs"}'
      creationTimestamp: null
      labels:
        app: test-app
        koordinator.sh/qosClass: LSR
    spec:
      containers:
      - args:
        - "3600"
        command:
        - sleep
        image: nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda10.2
        imagePullPolicy: IfNotPresent
        name: test
        resources:
          limits:
            cpu: "2"
            memory: 100Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      priorityClassName: koord-prod
      restartPolicy: Always
      schedulerName: koord-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
deployment/test-deploy created
```
4. Watch Pod status
```bash
$ kubectl get pod
NAME                           READY   STATUS    RESTARTS   AGE   IP          NODE                    NOMINATED NODE   READINESS GATES
test-deploy-7f9ddf4844-4qf6c   1/1     Running   0          35s   10.0.3.7    cn-beijing.10.0.3.245   <none>           <none>
test-deploy-7f9ddf4844-k4nws   1/1     Running   0          35s   10.0.3.10   cn-beijing.10.0.3.245   <none>           <none>
test-deploy-7f9ddf4844-ljfw4   1/1     Running   0          35s   10.0.3.6    cn-beijing.10.0.3.245   <none>           <none>
```

5. Get Reservation information from Pods, expect two Pods allocate CPU Core from Reservation
```bash
$ kubectl get pod -o yaml | grep reserva
      scheduling.koordinator.sh/reservation-allocated: '{"name":"test-gfr29","uid":"7da25eb5-8873-4e72-8451-0740aa247572"}'
      scheduling.koordinator.sh/reservation-allocated: '{"name":"test-gfr29","uid":"7da25eb5-8873-4e72-8451-0740aa247572"}'
```
6.  Verify Pods got CPU Cores, expect two Pod allocated from Reservation
```bash
 $ kubectl get pod -o yaml | grep cpuset
      scheduling.koordinator.sh/resource-status: '{"cpuset":"0,2"}'
      scheduling.koordinator.sh/resource-status: '{"cpuset":"4,6"}'
      scheduling.koordinator.sh/resource-status: '{"cpuset":"1,3"}'
```

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

**Comment by eahydra** [comment]
Implemented based on #1139. We can review the code of `pkg/scheduler/plugins/nodenumaresources` first, and then merge this after #1139 is merged.

**Review by FillZpp — APPROVED** [review]
/lgtm

**Pull request merged**
